server/src/main/java/org/elasticsearch/index/IndexModule.java

```diff
@@ -175,8 +175,8 @@ public Index getIndex() {
      *
      * @return the engine factory
      */
-    EngineFactory getEngineFactory() {
-        return engineFactory;
+    Collection<Function<IndexSettings, Optional<EngineFactory>>> engineFactoryProviders() {
```

Comment:
In which case are there multiple engineFactoryProviders? Isn't it sufficient to only add the indexSettings?


*Update*: Oh we had multiple engineFactories already before. All good - maybe worth a follow up to see if this is still required.

Reply:
Yes this was already like that but when resolving a factory the code bails out if multiple providers are set. This looks a bit weird, maybe related to the plugin logic. I'd follow up and change it to only one provider but this will break the plugin API. I think its fine for us.

Reply:
Or maybe worth doing this right now (as I'm already touching these parts)?

Reply:
I don't mind either way. If it is an easy change, fine to do it, otherwise I'd defer for later.

Reply:
@mfussenegger 

> In which case are there multiple engineFactoryProviders? Isn't it sufficient to only add the indexSettings?

Is your suggestion to only allow a single configured factory provider and such remove the `EnginePlugin` iface (which is the main reason to support a list of providers)?

Reply:
I initially thought this PR introduces the possibility to have multiple different EngineFactory /EnginePlugins active at the same time - then saw that it was already there before but didn't have a closer look at where it is coming from/why.

I just had a closer look now at the `EnginePlugin` usage and it seems the list is only their due to how the plugin discovery works. Doesn't look like it is worth to change it now.